dsmr/node_test.go

```diff
@@ -0,0 +1,1346 @@
+				{
+					txs: []tx{
+						{
+							ID:     ids.GenerateTestID(),
+							Expiry: 2,
+						},
+					},
+					expiry: 2,
+				},
+			},
+			parent: Block{
+				ParentID:  ids.GenerateTestID(),
+				Height:    1,
+				Timestamp: 1,
+				blkID:     ids.GenerateTestID(),
+			},
+			timestamp: 2,
+		},
+		{
+			name: "multiple chunk certs",
+			chunks: []chunk{
+				{
+					txs: []tx{
+						{
+							ID:     ids.GenerateTestID(),
+							Expiry: 2,
+						},
+					},
+					expiry: 2,
+				},
+				{
+					txs: []tx{
+						{
+							ID:     ids.GenerateTestID(),
+							Expiry: 2,
+						},
+					},
+					expiry: 2,
+				},
+				{
+					txs: []tx{
+						{
+							ID:     ids.GenerateTestID(),
+							Expiry: 2,
+						},
+					},
+					expiry: 2,
+				},
+			},
+			parent: Block{
+				ParentID:  ids.GenerateTestID(),
+				Height:    1,
+				Timestamp: 1,
+				blkID:     ids.GenerateTestID(),
+			},
+			timestamp: 2,
+		},
+		{
+			name: "one expired and one pending chunk cert",
+			chunks: []chunk{
+				{
+					txs: []tx{
+						{
+							ID:     ids.GenerateTestID(),
+							Expiry: 1,
+						},
+					},
+					expiry: 1,
+				},
+				{
+					txs: []tx{
+						{
+							ID:     ids.GenerateTestID(),
+							Expiry: 3,
+						},
+					},
+					expiry: 3,
+				},
+			},
+			parent: Block{
+				ParentID:  ids.GenerateTestID(),
+				Height:    1,
+				Timestamp: 1,
+				blkID:     ids.GenerateTestID(),
+			},
+			timestamp: 2,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := require.New(t)
+			sk, err := bls.NewSecretKey()
+			r.NoError(err)
+
+			node, err := New[tx](
+				ids.EmptyNodeID,
+				sk,
+				NoVerifier[tx]{},
+				p2ptest.NewClient(
+					t,
+					context.Background(),
+					&p2p.NoOpHandler{},
+					ids.EmptyNodeID,
+					ids.EmptyNodeID,
+				),
+				p2ptest.NewClient(
+					t,
+					context.Background(),
+					&p2p.NoOpHandler{},
+					ids.EmptyNodeID,
+					ids.EmptyNodeID,
+				),
+				p2ptest.NewClient(
+					t,
+					context.Background(),
+					&p2p.NoOpHandler{},
+					ids.EmptyNodeID,
+					ids.EmptyNodeID,
+				),
+				nil,
+			)
+			r.NoError(err)
+
+			wantChunks := make([]Chunk[tx], 0)
+			for _, chunk := range tt.chunks {
+				chunk, err := node.NewChunk(
+					context.Background(),
+					chunk.txs,
+					chunk.expiry,
+					chunk.beneficiary,
+				)
+				r.NoError(err)
+
+				// Only expect chunks that have not expired
+				if chunk.Expiry < tt.timestamp {
+					continue
+				}
+
+				wantChunks = append(wantChunks, chunk)
+			}
+
+			blk, err := node.NewBlock(tt.parent, tt.timestamp)
+			r.ErrorIs(err, tt.wantErr)
+			if err != nil {
+				return
+			}
+
+			r.Equal(tt.parent.GetID(), blk.ParentID)
+			r.Equal(tt.parent.Height+1, blk.Height)
+			r.Greater(blk.Timestamp, tt.parent.Timestamp)
+			packer := wrappers.Packer{Bytes: make([]byte, 0), MaxSize: MaxMessageSize}
+			r.NoError(codec.LinearCodec.MarshalInto(blk, &packer))
+			r.Equal(utils.ToID(packer.Bytes), blk.GetID())
```

Comment:
Is this check necessary? I'm not sure checking the block serialization is needed within this test

Reply:
In hindsight this is kind of testing implementation details. We can probably do something simpler like just check that the id field is populated...